### ISSUE TITLE
implement inventory methods and hooks

### ISSUE THREAD

**Issue by aaronjudd** [reactioncommerce/reaction#510]
The current inventory update mechanism is using **collection hooks**, and updates the product modifier based on a fairly rigid bit of logic that's tough to understand, and in a difficult location to easily extend or replace.

These changes will make it possible to use method hooks to transform inventory before or after product inventory methods as described in a workflow object. 

The addition of the workflow object to the Product collection will let us add a new publishing workflow (anyone with role x in workflows can see, role y can publish), in addition to providing some additional options for customization within the method calls. Publishing is not part of this feature, but some groundwork like this is.

## changes
- transition product/variant create defaults to the schema
- add workflow object to Products collection
- Remove existing collection hooks with a hook on product update that calls "inventory/adjust" and updates quantity of inventory records as needed.
- Add new hook to cart, that calls "inventory/addReserve" when item added to cart.
- Add clearReserve to cart item delete methods.
- Add Inventory collection document, proposed keys:
  ```
    id
    productId
    variantId
    workflow.status
    workflow.workflow
  ```

*Note: Could be expanded with fulfillment data, other metadata.*
- Add `common` inventory methods:

```
"inventory/register": function (product) {
  // adds or updates inventory collection with this product
  // and updates any inventory hierarchy
},

"inventory/adjust": function (product) {
  // return adjusted product object based on inventory hierarchy
  // add logic per product.type
},
"inventory/addReserve": function (productId, variantId, qty){
 // changes inventory workflow status to reserve/hold
},
"inventory/clearReserve": function (productId, variantId, qty) {
 // changes inventory workflow status to available
},
"inventory/lowStock": function (productId, variantId, qty) {
 // return stock status for product
 // send email / notifications on low stock
},
"inventory/shipped": function (productId, variantId, qty) {
 // updates inventory with shipped status
},
"inventory/return": function (productId, variantId, qty) {
  // updates inventory with return
}
```

## Potential additional refactors 

For product/variants we could remove `qty` from the document altogether, however I like the idea of self contained documents, and the easy tracking/reporting that will give us with revision control (among other things), so we'll use the `inventory/adjust` method to keep the collections synched from the `ReactionCore.Collections.Products.before.update` hook.

We could change the actual client helpers to use transforms, similar to the cart totals, to provide any kind of inventory transformations as well.

Product.type and product.variant.type could be used to optional workflow handling.

Inventory could be synched to cart documents as well, if we saw the need.

**Comment by spencern** [comment]
Our needs at @getoutfitted will require us to hook into inventory in a few different spots.

#### Rental Inventory Tracking
We need to track rental products on an item-by-item basis. 

A while back I wrote some code that added the idea of `inventory` variants for products - enabling shop owners to create 'inventory' variants for each unit of something they had in stock.

We need this concept of inventory variants so that we can track whether or not a unit (UID) is booked or available for a given set of dates. Potentially this type of inventory could be it's own collection or could be an array within the inventory collection. I'll have to think about exactly how it should function within this new inventory system. Currently we are using The Most Complicated Collection Hook Ever to check to see if anything is in stock and to 'reserve' a product on `order/inventoryAdjust` (not the same as the proposed reserve)

Not sure exactly how this should affect or how I would add on to the proposed inventory collection. Potentially a product of type rental could have multiple inventory documents for each variantId, but that seems like it could get complicated.

#### Traditional Inventory Tracking
We also sell products, so we need a way to track traditional inventory as well. I like the idea of using transforms similar to cart totals. Excited to get away from the collection hooks as they got really messy once we were more than once level deep in in the variants.

#### Adjustment on Checkout
We need to be able to hook into the inventory adjustment after a checkout and, if a product is a rental, update the dates that a UID is available for rental products.

#### Checking inventory availability
We need to be able to look up if/how many of an item is available based on the dates that a customer has chosen (we are currently storing those dates on the cart/order) and the availability calendar of the items they are browsing or adding to cart.

#### Cancelled, changed, or returned orders
We need some type of system (and are building out a CSR dashboard for our internal CSR staff) that will let authenticated users view and make changes to an existing order and update the inventory appropriately. I'd imagine that this is what `inventory/return` does in this proposal.

#### Inventory status change while in users cart
A wishlist item that I've had for a while and is a reason for using a reactive system would be if an items availability changes while it's in a cart or carts, that those users would be notified in real time that an item in their cart is now 'low stock' or 'out of stock' I know right now that the cart is cloned and there would have to be an additional lookup to update items in the cart, but I think it's worth discussing.

#### Other stuff.
I like the idea of `inventory/addReserve` - will a reserved item have a time-limit or I guess that's something that should be managed with `inventory/clearReserve`?
Interested in hearing more about this idea.

@paulgrever and I have discussed a few times the idea of having a Variants collection that would make variants a lot quicker to find and update. This Inventory collection might do enough to make the need for a variants collection unnecessary.

**Comment by spencern** [comment]
I'm working on a operations dashboard for our warehouse staff that will give them a heads up view of what our inventory status is for any given product (calendar view for us)

I'm looking for the best way to identify which variants of a product are actual products and should actually have inventory accounted for. I don't feel like I fully understand the best way to do this in reaction currently and perhaps this will play into #510 and how that is built out.

This becomes more complicated the more options you have on any variant, but let's take a T-Shirt that comes in multiple colors and sizes as an example.

In Reaction Commerce this item is setup in the following way. (note: you might setup the sleeve length variant as a different product in real life, but it helps to illustrate the point)
```
Product - T Shirt
  - variant Red (Inv Qty = sum of children, but also we don't care)
    - variant Large (Inv Qty = sum of children, but also we don't care)
      - variant Long Sleeves (qty = 3) (Customer facing name is Red Large Long Sleeve)
      - variant Short Sleeves (qty = 4)
    - variant Medium (qty = 1 ) (Customer facing: Red Medium)
    - variant Small (qty = 2)
  - variant Green (Inv Qty = sum of children, but also we don't care)
    - variant Large (qty)
    - variant Medium (qty)
    - variant Small (qty)
  ```
In this example, the only variants that are 'real' are the child-most variants (variants without children themselves). In Reaction - these variants are listed flat in the products `variants` array and the only way to traverse them or distinguish them is to find other variants which have a parent ID that is equal to the current variantID you are looking for. What I'm trying to get at, is that we don't really care that much about how many qty of a particular parent variant we have
  
The way I understand it, a better way to visualize this be like so:
  
| Title   | Option 1 - Color | Option 2 - Size | Option 3 - Sleeve Length | Qty |
|---------|------------------|-----------------|--------------------------|----|
| T Shirt | Red              | Large           | Long Sleeve              | 3 |
| T Shirt | Red              | Large           | Short Sleeve             | 2 |
| T Shirt | Red              | Medium       | (default)                          | 3 |
| T Shirt | Red              | Small           | (default)                         | 3 |
| T Shirt | Green            | Large         |  (default)                        | 3 |
| T Shirt | Green            | Medium       | (default)                         | 2 |
| T Shirt | Green            | Small           | (default)                         | 17 |

That is to say, there is no such product as a 'Red' T Shirt - and there is no such product as a 'Red Large' T Shirt, but only once you define all of the possible options does the variant make sense. E.g. 
'Red Large Long Sleeve' T Shirt

#### The way I currently do it
The way I currently figure out what these 'real' variants that should actually have inventory are in Reaction is to find within a product - 
```
all `variant._id`'s for which there are no variants that have that `variant._id` as a `parentId`
```

If a product has inventory variants (something we are discussing in #510 and is currently how we track rental products) then the pseudo query becomes 
```
all `variant._id`'s for which there are variants with `type: inventory` that have that `variant._id` as a `parentId`
```

Now that I've written all of this, it seems like the best way to deal with this might be to add a SKU to all of these 'real' variants.

1. Is there a better way for me to do this? Or is this just complicated?
2. It seems to be fairly expensive to do any lookups on variants as you've got to traverse the full variant parent tree to get a full understanding. (e.g. if the variant you are looking at is 'Long Sleeve', you have to go to it's parent to find that it's a 'Large' and again to find that it's 'Red'
3. We are discussing adding a 'Variants' collection which indexes and mirrors variants that are on products but makes it cheaper to lookup an individual variant. I'd imagine that you've thought about this as well at some point, so if you've got any insights there, I'm interested in that as well.

A bit of a ramble but wanted to get these thoughts out as they are things that we are dealing with thinking through on a regular basis when working with variants from non-product pages (e.g. orders, fulfillment, returns, processing, etc) See [reaction-advanced-fulfillment](https://github.com/getoutfitted/reaction-advanced-fulfillment) for example.

**Comment by aaronjudd** [comment]
I've been puzzling over this a bit myself, and I think the answer might be quite simple.  We currently update the inventory totals of the parent to be a total of the child variants -> but if we just stop doing that at least in the new inventory code, and just hide the UI elements (because the reason for that was purely UI driven), then we don't really have a complicated process -> IF the item has inventory in the inventory table... filter by status, and that's the end of it.   If it has 0/null qty, then it's not an inventory item.
